source/common/network/socket_interface_impl.cc

```diff
@@ -81,6 +82,18 @@ bool SocketInterfaceImpl::ipFamilySupported(int domain) {
   return SOCKET_VALID(result.rc_);
 }
 
+Server::BootstrapExtensionPtr
+SocketInterfaceImpl::createBootstrapExtension(const Protobuf::Message&,
+                                              Server::Configuration::ServerFactoryContext&) {
+  return std::make_unique<SocketInterfaceExtension>(this);
+}
+
+ProtobufTypes::MessagePtr SocketInterfaceImpl::createEmptyConfigProto() {
+  return std::make_unique<envoy::config::core::v3::DefaultSocketInterface>();
+}
+
+REGISTER_FACTORY(SocketInterfaceImpl, Server::Configuration::BootstrapExtensionFactory);
+
 static SocketInterfaceLoader* socket_interface_ =
```

Comment:
Shall we do this in `server.cc`? i.e. if another socket interface is registered and specified as default, we don't need to instantiate default socket interface?

Reply:
Definitely a good point but if tests are run without the server being properly initialized, the `SocketInterfaceSingleton` won't be properly initialized. Just ran a quick test and they do crash.

Should we maybe statically initialize the singleton to the default `SocketInterface` in some common test file, like "test/test_common/environment.cc"?

Reply:
Turns out integration:hotrestart_test, integration:run_envoy_test and test:router_tool_test still fail :-(

Reply:
> Should we maybe statically initialize the singleton to the default `SocketInterface` in some common test file, like `test/test_common/environment.cc`?

Yeah I think that's good way to do test.

OTOH, if this causes a lot churn, we can do static registration here, then let's not do factory registration for default interface. Because even they're configured through `bootstrap_extensions`, it has nothing different than statically registered one?

Reply:
> > Should we maybe statically initialize the singleton to the default `SocketInterface` in some common test file, like `test/test_common/environment.cc`?
> 
> Yeah I think that's good way to do test.

Because we do some network related things early on, before initializing the server, (e.g., call `validateIpv4Supported`), we'll have to initialize very early on the `SocketInterfaceSingleton` to something. 

> 
> OTOH, if this causes a lot churn, we can do static registration here, then let's not do factory registration for default interface. Because even they're configured through `bootstrap_extensions`, it has nothing different than statically registered one?

Do you mean to keep the global static registration in `socket_interface_impl.cc` (which is a bit of a misnomer at this point but that's for another time)? Or do you think we should move that to `server.cc` (and keep it as a global static)? I'm fine either way. 

As for the factory registration, here's the weird use case I came up with: people may want to configure a different default but may still want to configure addresses to use what we now call `SocketInterfaceImpl`, which is registered with name `envoy.config.core.default_socket_interface`. So, we still end up needing the factory registration for name lookups, unless we decide we don't support that. 

Tying this to your first suggestion:  we could maybe do a `SocketInterfaceSingleton::initialize(Network::socketInterface("envoy.config.core.default_socket_interface"))` very early on, e.g., somewhere in `main_common.cc` or in `PlatformImpl()` and that would avoid the need for the static init and allocating a redundant `SocketInterfaceImpl` unique ptr by reusing the factory registration.

Reply:
> > > Should we maybe statically initialize the singleton to the default `SocketInterface` in some common test file, like `test/test_common/environment.cc`?
> > 
> > 
> > Yeah I think that's good way to do test.
> 
> Because we do some network related things early on, before initializing the server, (e.g., call `validateIpv4Supported`), we'll have to initialize very early on the `SocketInterfaceSingleton` to something.
> 
> > OTOH, if this causes a lot churn, we can do static registration here, then let's not do factory registration for default interface. Because even they're configured through `bootstrap_extensions`, it has nothing different than statically registered one?
> 
> Do you mean to keep the global static registration in `socket_interface_impl.cc` (which is a bit of a misnomer at this point but that's for another time)? Or do you think we should move that to `server.cc` (and keep it as a global static)? I'm fine either way.

Either 
a. move initialization to `server.cc` and fix test
b. keep global static registration without declaring factory

works for me.

In long term we might want a. but to keep PR smaller and easier to review we can do a. as a follow up.

> 
> As for the factory registration, here's the weird use case I came up with: people may want to configure a different default but may still want to configure addresses to use what we now call `SocketInterfaceImpl`, which is registered with name `envoy.config.core.default_socket_interface`. So, we still end up needing the factory registration for name lookups, unless we decide we don't support that.

I meant we should drop it until we can support it in a meaningful way? i.e. when address can specify socket interface.

Reply:
If it's fine with you, let's try to do a. as a follow up after we fix `Address`.  If that's okay, I'll push a quick fix to the PR (`SocketInterface` lookup needs some casting).

Reply:
Not to forget about it, I pushed the fix now, but that's independent of the changes we've been discussing.

Reply:
sgtm

Reply:
Thanks! Let me know if anything else is needed (the update seems to have verified fine).